### PR TITLE
modernize cmake

### PR DESCRIPTION
**Pull request by ClausKlein**
add missing alias and export it
add .clang-format config file
apply cmake-format to cmake files
fix version used in cmake config package
add usable ctest configuration

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -16,26 +26,21 @@ option(USE_BOOST_REGEX "Replace std::regex with Boost.Regex" OFF)
 # C++ standard
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS OFF)
-if(NOT CMAKE_CXX_STANDARD OR CMAKE_CXX_STANDARD LESS 11)
-	set(CMAKE_CXX_STANDARD 11)
+if(NOT CMAKE_CXX_STANDARD OR CMAKE_CXX_STANDARD LESS 17)
+    set(CMAKE_CXX_STANDARD 17)
```
Comment (thelema) [inline CMakeLists.txt:30]: 1) if you're modernizing, this needs some rewriting. This isn't modern cmake as it's global instead of target specific, no?
2) While switching to c++17 might be useful for some users, mixing it with this pull request makes it hard to accept the rest of the benefits of this change, as dropping all support for older compilers is probably not a desirable change.

File: CMakeLists.txt [hunk 2]
```diff
@@ -62,6 +62,8 @@ endif()
 target_include_directories(docopt PUBLIC $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}> $<INSTALL_INTERFACE:include/docopt>)
 target_include_directories(docopt_s PUBLIC $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}> $<INSTALL_INTERFACE:include/docopt>)
 
+include(cmake/WarningsAsErrors.cmake)
```
Comment (thelema) [inline CMakeLists.txt:65]: This is way more than fixing the project name.  I'm suspicious that -Werror is the right thing to include as part of a cmake cleanup pull request. I'd want to see CI tests passing after this change before accepting this change.

File: CMakeLists.txt [hunk 3]
```diff
@@ -16,26 +26,21 @@ option(USE_BOOST_REGEX "Replace std::regex with Boost.Regex" OFF)
...
-		docopt_value.h
-		)
+set(docopt_HEADERS docopt.h docopt_private.h docopt_util.h docopt_value.h)
 
 #============================================================================
 # Compile targets
 #============================================================================
 if(MSVC OR XCODE)
-    # MSVC requires __declspec() attributes, which are achieved via the 
+    # MSVC requires __declspec() attributes, which are achieved via the
```
Comment (thelema) [inline CMakeLists.txt:43]: Unnecessary whitespace change; remove

File: CMakeLists.txt [hunk 4]
```diff
@@ -50,65 +55,64 @@ else()
     set_target_properties(docopt_o PROPERTIES POSITION_INDEPENDENT_CODE TRUE)
 
     add_library(docopt SHARED $<TARGET_OBJECTS:docopt_o>)
-	set_target_properties(docopt PROPERTIES
-			VERSION ${PROJECT_VERSION}
-			SOVERSION ${PROJECT_VERSION_MAJOR}
-			)
+    set_target_properties(docopt PROPERTIES VERSION ${PROJECT_VERSION} SOVERSION ${PROJECT_VERSION_MAJOR})
```
Comment (thelema) [inline CMakeLists.txt:58]: Unnecessary whitespace change; remove.

File: CMakeLists.txt [hunk 5]
```diff
@@ -16,26 +26,21 @@ option(USE_BOOST_REGEX "Replace std::regex with Boost.Regex" OFF)
...
 # Sources & headers
 #============================================================================
 set(docopt_SOURCES docopt.cpp)
-set(docopt_HEADERS
-		docopt.h
-		docopt_private.h
-		docopt_util.h
-		docopt_value.h
-		)
+set(docopt_HEADERS docopt.h docopt_private.h docopt_util.h docopt_value.h)
```
Comment (thelema) [inline CMakeLists.txt:37]: Unnecessary whitespace change; remove

File: CMakeLists.txt [hunk 6]
```diff
@@ -50,65 +55,64 @@ else()
...
-	# This is needed on Linux, where linking a static library into docopt.so
-	# fails because boost static libs are not compiled with -fPIC
-	set(Boost_USE_STATIC_LIBS OFF)
-    find_package(Boost 1.53 REQUIRED COMPONENTS regex)
-    include_directories(${Boost_INCLUDE_DIRS})
-    target_link_libraries(docopt ${Boost_LIBRARIES})
-	if(WITH_STATIC)
-		target_link_libraries(docopt_s ${Boost_LIBRARIES})
-	endif()
+    # This is needed on Linux, where linking a static library into docopt.so
```
Comment (thelema) [inline CMakeLists.txt:88]: This is a good modernization, switching from add_definitions to target_compile_definitions on specific projects

File: CMakeLists.txt (hunk 1 above)
Comment (ClausKlein) [reply CMakeLists.txt:29]: The current standard is c++20
c++14 is deprecated
and c++11 is buggy

File: CMakeLists.txt (hunk 5 above)
Comment (ClausKlein) [reply CMakeLists.txt:36]: I use `cmake-format -i CMakeLists.txt` to format the hole file!

**Comment by ClausKlein** [comment]
> Can we rebase this on top of tip in order to find what's still outstanding?

done

File: CMakeLists.txt (hunk 2 above)
Comment (ClausKlein) [reply CMakeLists.txt:51]: what is the problem with this way to be correct while building this lib?

File: CMakeLists.txt (hunk 5 above)
Comment (thelema) [reply CMakeLists.txt:36]: Ah, I misunderstood that.  I think that collapsing the header list is a pessimization and I appreciate filelists like this being one line per file to make diffs simple.

Comment (ClausKlein) [reply CMakeLists.txt]: done

File: CMakeLists.txt (hunk 2 above)
Comment (ClausKlein) [reply CMakeLists.txt:57]: There are no warnings and too no errors, luckily.

File: CMakeLists.txt [hunk 7]
```diff
@@ -16,73 +25,90 @@ option(USE_BOOST_REGEX "Replace std::regex with Boost.Regex" OFF)
...
 endif()
 
 #============================================================================
 # Examples
 #============================================================================
 if(WITH_EXAMPLE)
-	add_executable(docopt_example examples/naval_fate.cpp)
-	target_link_libraries(docopt_example docopt)
+    add_executable(docopt_example examples/naval_fate.cpp)
+    target_link_libraries(docopt_example docopt::docopt)
```
Comment (stephan-cr) [inline CMakeLists.txt:93]:
```suggestion
    target_link_libraries(docopt_example PRIVATE docopt::docopt)
```

File: CMakeLists.txt [hunk 8]
```diff
@@ -16,73 +25,90 @@ option(USE_BOOST_REGEX "Replace std::regex with Boost.Regex" OFF)
...
-			ESCAPE_QUOTES
-	)
-	add_test("Testcases docopt" ${TESTPROG})
+    enable_testing()
+
+    find_program(PYTHON python REQUIRED)
+    set(TESTPROG "${CMAKE_CURRENT_BINARY_DIR}/run_testcase")
+    set(TESTCASES "${PROJECT_SOURCE_DIR}/testcases.docopt")
+    add_executable(run_testcase run_testcase.cpp)
+    target_link_libraries(run_testcase docopt::docopt)
```
Comment (stephan-cr) [inline CMakeLists.txt:106]:
```suggestion
    target_link_libraries(run_testcase PRIVATE docopt::docopt)
```

File: CMakeLists.txt (hunk 1 above)
Comment (jaredgrubb) [reply CMakeLists.txt:29]: I just saw this .. this library is intended to work with C++11 and has no C++17 features.

What do you mean by "C++11" is buggy?

File: cmake/WarningsAsErrors.cmake [hunk 9]
```diff
@@ -0,0 +1,31 @@
+if(CMAKE_CXX_COMPILER_ID MATCHES "Clang" AND CMAKE_CXX_SIMULATE_ID MATCHES
+                                             "MSVC"
+)
+    target_compile_options(${PROJECT_NAME} PRIVATE /W4 /WX)
+elseif(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
+    target_compile_options(
+        ${PROJECT_NAME}
+        PRIVATE -Wall
```
Comment (jaredgrubb) [inline cmake/WarningsAsErrors.cmake:8]: FWIW, I think we should _either_ have -Wall _or_ -Werror ... having both on puts the compilation at the mercy of the compiler version, as GCC 27 might add a brand new warning and suddenly everyone's build starts breaking.

If we have -Wall, then we should not have -Werror except for those warnings that we have explicitly fixed and promised do not exist.

File: CMakeLists.txt (hunk 1 above)
Comment (jaredgrubb) [reply CMakeLists.txt:29]: Maybe we should just remove the default completely and let it be "whatever the compiler wants to be default", and then we can have our CI system explicitly set it to other values (including 11 and 23)?

**Comment by ClausKlein** [comment]
you may use it or not it, it is your choice!
